Add parser for LegionBoard

AltNico opened the pull request:
I started working on the parser for [LegionBoard](https://legionboard.github.io/). As @johan12345 said, I made a copy of `SVPlanParser` to create [`LegionBoardParser`](https://github.com/johan12345/substitution-schedule-parser/compare/master...AltNico:addLegionBoard?expand=1#diff-a6e11366a1119512f844cd6eb7dd8245R1). Also, I scanned through [`SubstitutionScheduleData`](https://github.com/johan12345/substitution-schedule-parser/compare/master...AltNico:addLegionBoard?expand=1#diff-2d1ab1fb4ff397e4c8cee89e56961922R75) and  [`BaseParser`](https://github.com/johan12345/substitution-schedule-parser/compare/master...AltNico:addLegionBoard?expand=1#diff-8905e332d3ac9cf936a74ae870b13c2fR135) and added a case for `legionboard`.

@johan12345: Can you look through the changes to see whether I'm on the right way? I left a [`QUESTION` comment in `getSubstitutionSchedule()`](https://github.com/johan12345/substitution-schedule-parser/compare/master...AltNico:addLegionBoard?expand=1#diff-a6e11366a1119512f844cd6eb7dd8245R60) which you maybe can answer. Also, if you have any ideas related to a `TODO`, feel free to answer too :D

Related to #5.

AltNico commented:
#### TODO

[x] [for-each not applicable for `JSONArray`](https://travis-ci.org/johan12345/substitution-schedule-parser/builds/146320654#L156)

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+		 * QUESTION: Where do I implement the SHA-256 hash of "[username]//[password]"?
+		 * See the documentation: https://gitlab.com/legionboard/heart/blob/master/doc/README.md
+		 */
+		authKey = getCredential().getHash();
```
For authenticating to the API, you should create an own hashing implementation in this class (such as a `private String getHash(UserPasswordCredential credential)` function), `Credential.getHash()` is meant for a different use case.

You can retrieve the username and password by casting the credential to a `UserPasswordCredential`:

``` java
    UserPasswordCredential credential = (UserPasswordCredential) credential;
    String username = credential.getUsername();
    String password = credential.getPassword();
```

For implementing the SHA-256 hash, you can basically use the same code as in https://github.com/johan12345/substitution-schedule-parser/blob/master/parser/src/main/java/me/vertretungsplan/objects/credential/BaseCredential.java#L49, just inserting `username + "//" + password` as the data.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	private JSONArray getJSONArray(String url) throws IOException, JSONException {
+		// TODO: get JSONArray from HTML
```
This should basically be `return new JSONArray(httpGet(url, "UTF-8"))`.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+		List<String> teachers = new ArrayList<>();
+		final JSONArray JSONTeachers = getTeachers();
+		for (JSONArray teacher : JSONTeachers) {
+			teachers.add(teacher["name"]);
```
Indeed, it is unfortunately not possible to use a for-each on a JSONArray. You need to use

``` java
    for (int i = 0; i < JSONTeachers.length(); i++) {
        JSONObject teacher = teachers.getJSONObject(i);
        teachers.add(teacher.getString("name"))
    }
```

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+	@Override
+	public List<String> getAllTeachers() throws IOException, JSONException {
+		List<String> teachers = new ArrayList<>();
+		final JSONArray JSONTeachers = getTeachers();
```
just a quick reminder: Variables should start with lowercase letters in Java (see e.g. https://google.github.io/styleguide/javaguide.html#s5.2.7-local-variable-names) -> `jsonTeachers` (or `teachersJson`). :wink:

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+		List<String> classes = new ArrayList<>();
+		final JSONArray courses = getCourses();
+		for (JSONArray course : courses) {
+			classes.add(course["name"]);
```
see https://github.com/johan12345/substitution-schedule-parser/pull/6/files#r71672025

johan12345 commented:
Very nice that you have started to work on this! As you can see, I added some notes above. Feel free to ask if you have more questions.

AltNico commented:
@johan12345 Thank you for your feedback, I already pushed some commits with small fixes. :+1:

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	public SubstitutionSchedule getSubstitutionSchedule() throws IOException, JSONException, CredentialInvalidException {
+		api = data.getString("api");
```
I guess this line should be moved to the end of the constructor so that the `getAllClasses()` and `getAllTeachers()` methods can also be used without calling `getSubstitutionSchedule()` first.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,136 @@
+		final JSONArray courses = getCourses();
+		for (int i = 0; i < courses.length(); i++) {
+			JSONObject course = courses.getJSONObject(i);
+			classes.add(courses.getString("name"));
```
Typo: `courses.getString("name")` -> `course.getString("name")`

AltNico commented:
Yey, all checks have passed :dancers: 

#### TODO
- Implement `getHash(UserPasswordCredential credential)`
- Implement `parseLegionBoard(JSONArray changes, JSONArray courses, JSONArray teachers)`

AltNico commented:
Okay, so I implemented `getAuthenticationKey(UserPasswordCredential credential)`.

Can you tell me where I can pass the username and password? In the sample? I already modified it for LegionBoard's needs (Data, AutenticationData, ...).

johan12345 commented:
You just go to line [58](https://github.com/johan12345/substitution-schedule-parser/blob/master/sample/src/main/java/me/vertretungsplan/sample/Sample.java#L58) and change that call to

``` java
    UserPasswordCredential credential = new UserPasswordCredential();
    credential.setUsername("username");
    credential.setPassword("password");
    SubstitutionSchedule schedule = ParserUtil.parseSubstitutionSchedule(data, credential);
```

AltNico commented:
@johan12345 Thanks, I got it working. However, I had other problems because `getCredential()` does not work in the constructor and the `SHA-256` hasher you use hashes something different. I compared it with duckduckgo's sha256 hash and the resulting strings were different...

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -85,15 +83,7 @@ private String getAuthenticationKey(Credential credential) {
-			e.printStackTrace();
-		}
-		return null;
+		return DigestUtils.sha256Hex(username.toLowerCase() + "//" + password);
```
Ah, the difference is that mine uses Base64 encoding instead of the (in this context) more widely-used hexadecimal encoding. The advantage of Base64 is that it saves space while still being URL-safe, but if LegionBoard uses hexadecimal, you obviously need to use that here too :wink:

AltNico replied:
Darauf hab ich gar net geachtet o_O
Danke für die Erklärung!

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -119,7 +120,12 @@ private JSONArray getTeachers() throws IOException, JSONException {
+		try {
+			return new JSONArray(httpGet(url, "UTF-8"));
+		} catch (HttpResponseException e) {
+			e.printStackTrace();
```
Well, even though I (and many other people) do it often, just catching exceptions and printing their stack trace is a bad habit, especially if you actually expect these exceptions to occur sometimes (such as when the server is down). If you need no special handling of HTTP error codes, I would just remove the try-catch block and let it throw the `HttpResponseException` (which is a subclass of `IOException`).

AltNico replied:
Letting the function throwing the exception did not work for me because then the execution stopped.

johan12345 replied:
Yes, but it should stop if there is an error, shouldn't it?

AltNico replied:
Teachers and courses can be empty and there can still be changes, as these two fields are not required for a change. I'm currently building `parseLegionBoard()` and there I'm checking whether the changes are empty. But having empty teachers and courses is no problem.

Maybe I should try and catch in getChanges/Courses/... and stop execution there? But this was the easiest solution :P

johan12345 replied:
Ah, okay, die API gibt `404` zurück wenn keine Lehrer bzw. Kurse existieren. Dann würde ich die Exception erst in getAllClasses und getAllTeachers fangen, aber dann nicht `e.printStackTrace()` sondern prüfen, was der Fehlercode ist und bei `404` direkt `return null` und sonst mit `throw e` die Exception weiterreichen (bzw. letztendlich anhalten).

AltNico replied:
Ach komm, ich wollte noch'n bisschen mein Englisch verbessern ;D

OK, mach ich so ... irgendwann :P

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -61,10 +61,10 @@ public LegionBoardParser(SubstitutionScheduleData scheduleData, CookieProvider c
 	}
 
 	public SubstitutionSchedule getSubstitutionSchedule() throws IOException, JSONException, CredentialInvalidException {
+		authKey = getAuthenticationKey(getCredential());
```
Ah, that's true, it doesn't work in the constructor because the credential is set using `setCredential(Credential credential)` afterwards. To still make it work when `getAllClasses` (or Teachers) is called independently, you could also just drop the `authKey` field and just use `getAuthenticationKey` everywhere. Or override the `setCredential(Credential credential)` method to update the `authKey` field there.

AltNico commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,197 @@
+			return getJSONArray(url);
+		} catch (HttpResponseException e) {
+			// TODO: Implement
+			final boolean is404 = true;
```
@johan12345 Hier, hab's erstmal provisorisch gemacht.

johan12345 replied:
@AltNico Wieso nicht einfach gleich `if (e.getStatusCode() == 404)`?

AltNico replied:
Erledigt. Ich hab mir gestern nicht mehr die Doku angeguckt.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,197 @@
+
+	private void parseLegionBoard(SubstitutionSchedule substitutionSchedule, JSONArray changes, JSONArray courses, JSONArray teachers) throws IOException, JSONException {
+		if (changes == null) {
+			throw new IOException("LegionBoard has no changes.");
```
Ein leerer Vertretungsplan sollte definitiv keine Exception geben, sondern einen leeren `SubstitutionSchedule`.  Bei den anderen Parsern sind da dann noch die jeweiligen Tage mit drin (also wenn die Pläne für heute und morgen online sind, einfach zwei `SubstitutionScheduleDays` ohne `Substitutions`). Das wird hier wahrscheinlich schwierig, oder?

AltNico replied:
Exception wird nicht mehr geworfen.
Standardmäßig werden in LegionBoard Eye (Webseite) die Änderungen von 
heute (Alias now) bis in einer Woche (Alias i1w/in 1 week) abgerufen. 
Wie sieht es bei Dir normal aus?

johan12345 replied:
Das ist je nach Schule unterschiedlich. Bei den meisten gibt es nur nur heute und morgen, bei manchen sogar nur heute. Bei manchen (mit dem Untis-Infostundenplan zum Beispiel) aber auch ~1 Woche. Wenn bei Eye eine Woche der Standard ist, würde ich das hier auch so machen.

AltNico replied:
Kann man in der App die Änderungen nach Datum filtern? Heart kann nämlich theoretisch alle Änderungen, die jemals erstellt wurden, anzeigen. Wenn das andere Systeme nicht können (ich kenn mich da nicht aus), dann kann das die App wohl eher nicht, oder?

On July 23, 2016 1:41:49 PM GMT+02:00, Johan von Forstner notifications@github.com wrote:

> > -            // TODO: Implement
> > -            final boolean is404 = true;
> > -            if (is404) {
> > -                return null;
> > -            }
> > -            throw e;
> > -        }
> > -    }
> >   +
> > -    private JSONArray getJSONArray(String url) throws IOException,
> >   JSONException, HttpResponseException {
> > -        return new JSONArray(httpGet(url, "UTF-8"));
> > -    }
> >   +
> > -    private void parseLegionBoard(SubstitutionSchedule
> >   substitutionSchedule, JSONArray changes, JSONArray courses, JSONArray
> >   teachers) throws IOException, JSONException {
> > -        if (changes == null) {
> > -            throw new IOException("LegionBoard has no changes.");
> 
> Das ist je nach Schule unterschiedlich. Bei den meisten gibt es nur nur
> heute und morgen, bei manchen sogar nur heute. Bei manchen (mit dem
> Untis-Infostundenplan zum Beispiel) aber auch ~1 Woche. Wenn bei Eye
> eine Woche der Standard ist, würde ich das hier auch so machen.
> 
> ---
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/johan12345/substitution-schedule-parser/pull/6/files/c18972380dbbcfa4d07b02a7366a6277712bc391#r71971404

johan12345 replied:
Nein, die App zeigt immer alle Tage an, die sie vom Parser bekommt.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,203 @@
+		// Link course IDs to their names
+		HashMap<String, String> coursesHashMap;
+		if (courses != null) {
+			coursesHashMap = new HashMap<>();
```
You also need to initialize `coursesHashMap` if `courses == null`, otherwise using it below makes the build fail. Same for `teachersHashMap`.

AltNico replied:
Ich will später eine Exception werfen, wenn eine Änderung einen 
Teacher/Course hat, aber die Hashmap mit Teachers/Courses leer ist.

johan12345 replied:
Ah, stimmt, das ist sinnvoll. Dann musst du aber `coursesHashMap` auf `null` setzen, sonst kannst du das unten auch nicht prüfen :wink:

AltNico replied:
Wenn ich `coursesHashMap` nur nach der `if courses not null` Abfrage inintalisiere und fülle, ist es doch standardmäßig null, oder?

On July 24, 2016 11:37:36 AM GMT+02:00, Johan von Forstner notifications@github.com wrote:

> > -        } catch (HttpResponseException httpResponseException) {
> > -            if (httpResponseException.getStatusCode() == 404) {
> > -                return null;
> > -            }
> > -            throw httpResponseException;
> > -        }
> > -    }
> >   +
> > -    private void parseLegionBoard(SubstitutionSchedule
> >   substitutionSchedule, JSONArray changes, JSONArray courses, JSONArray
> >   teachers) throws IOException, JSONException {
> > -        if (changes == null) {
> > -            return;
> > -        }
> > -        // Link course IDs to their names
> > -        HashMap<String, String> coursesHashMap;
> > -        if (courses != null) {
> > -            coursesHashMap = new HashMap<>();
> 
> Ah, stimmt, das ist sinnvoll. Dann musst du aber `coursesHashMap` auf
> `null` setzen, sonst kannst du das unten auch nicht prüfen :wink: 
> 
> ---
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/johan12345/substitution-schedule-parser/pull/6/files/87f679af0282a20db485917fde2201644a993d7e#r71985493

johan12345 replied:
Nein, so was gibt es nur bei primitiven Typen (bei `boolean` `false` und bei `int` `0` etc.). So gibt es halt einen Compiler-Fehler sobald du versuchst, an einer Stelle darauf zuzugreifen, an der es eventuell nicht initialisiert wurde.

AltNico replied:
Ah, danke für die Erklärung. Ich bin wohl schon zu verwöhnt von PHP bei Heart :D

On July 24, 2016 11:54:18 AM GMT+02:00, Johan von Forstner notifications@github.com wrote:

> > -        } catch (HttpResponseException httpResponseException) {
> > -            if (httpResponseException.getStatusCode() == 404) {
> > -                return null;
> > -            }
> > -            throw httpResponseException;
> > -        }
> > -    }
> >   +
> > -    private void parseLegionBoard(SubstitutionSchedule
> >   substitutionSchedule, JSONArray changes, JSONArray courses, JSONArray
> >   teachers) throws IOException, JSONException {
> > -        if (changes == null) {
> > -            return;
> > -        }
> > -        // Link course IDs to their names
> > -        HashMap<String, String> coursesHashMap;
> > -        if (courses != null) {
> > -            coursesHashMap = new HashMap<>();
> 
> Nein, so was gibt es nur bei primitiven Typen (bei `boolean` `false`
> und bei `int` `0` etc.). So gibt es halt einen Compiler-Fehler sobald
> du versuchst, an einer Stelle darauf zuzugreifen wo es eventuell nicht
> initialisiert wurde.
> 
> ---
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/johan12345/substitution-schedule-parser/pull/6/files/87f679af0282a20db485917fde2201644a993d7e#r71985718

AltNico commented:
@johan12345: Can you tell me whether I understood implementing it right?

I have to add `SubstitutionScheduleDay`s to the `SubstitutionSchedule`. This `SubstitutionScheduleDay`s contain several 'Substitution `s.

Now some more questions:
- LegionBoard has no 'Last time changed'. Can I leave this field empty or should I set it to the current time?
- Change A goes from day 1 to day 3. Should I create three `SubstitutionScheduleDay`s with one `Substitution` each?
- Do I have to do anything in order for filtering to work? For example overriding the functions?

johan12345 commented:
> I have to add SubstitutionScheduleDays to the SubstitutionSchedule. This SubstitutionScheduleDays contain several 'Substitution `s.

Correct.

> LegionBoard has no 'Last time changed'. Can I leave this field empty or should I set it to the current time?

I think it should work if you leave it empty.

> Change A goes from day 1 to day 3. Should I create three SubstitutionScheduleDays with one Substitution each?

Yes, I think that would be the best way as the data structure has no other way to represent things that span multiple days. But do you know of an example where this would be relevant? Is it only used for things like holidays?

> Do I have to do anything in order for filtering to work? For example overriding the functions?

No, filtering should work out of the box when the appropriate fields (classes/teachers/subjects) are set.

AltNico commented:
@johan12345: If teachers are ill, there is often a cancellation for the whole week. On our school we often have multi-day changes.

On July 31, 2016 5:57:25 PM GMT+02:00, Johan von Forstner notifications@github.com wrote:

> > I have to add SubstitutionScheduleDays to the SubstitutionSchedule.
> > This SubstitutionScheduleDays contain several 'Substitution `s.
> 
> Correct.
> 
> > LegionBoard has no 'Last time changed'. Can I leave this field empty
> > or should I set it to the current time?
> 
> I think it should work if you leave it empty.
> 
> > Change A goes from day 1 to day 3. Should I create three
> > SubstitutionScheduleDays with one Substitution each?
> 
> Yes, I think that would be the best way as the data structure has no
> other way to represent things that span multiple days. But do you know
> of an example where this would be relevant? Is it only used for things
> like holidays?
> 
> > Do I have to do anything in order for filtering to work? For example
> > overriding the functions?
> 
> No, filtering should work out of the box when the appropriate fields
> (classes/teachers/subjects) are set.
> 
> ---
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/johan12345/substitution-schedule-parser/pull/6#issuecomment-236437965

johan12345 commented:
Okay, that makes sense if the changes are not input for each lesson where the teacher is ill but instead for the whole time period. I think adding it to each `SubstitutionScheduleDay` is still the only reasonable option.

AltNico commented:
OK, so this is nearly done. The parser is working and I also modified the sample to run with a LegionBoard Heart instance. The last thing to do is [multi-day changes support](https://github.com/johan12345/substitution-schedule-parser/pull/6/files#diff-a6e11366a1119512f844cd6eb7dd8245R178).

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,327 @@
+				if (!startingHour.equals("") && !endingHour.equals("")) {
+					lesson = startingHour + ". - " + endingHour + ". Std.";
+				}
+				substitution.setLesson(lesson);
```
The app displays the lesson in quite a large font, so it's probably better to just set the lesson to something like `1` (for single lessons) or `1 - 2` (for multiple lessons).

AltNico replied:
OK, agreed. But what do we do with "Ab 3. Stunde" and "Bis 6. Stunde"?

johan12345 replied:
Well, maybe "Ab 3" and "1 -  6"?

AltNico replied:
Done.

AltNico committed:
Add parser for LegionBoard

Fixes #5.

AltNico committed:
Add LegionBoard to supported systems

AltNico committed:
Add case for LegionBoard

AltNico commented:
Ready for review and merge :D

johan12345 commented:
@AltNico Could you give me a URL and credentials to a LegionBoard instance so that I can test the parser before merging it? (see also my email)

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+		return getJSONArray(url);
+	}
+
+	private JSONArray getJSONArray(String url) throws IOException, JSONException, HttpResponseException {
```
You can remove the `HttpResponseException` from the `throws` list because it is a subclass of `IOException`.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+		for (int i = 0; i < jsonArray.length(); i++) {
+			sortedJsonArray.put(jsonValues.get(i));
+		}
+		return sortedJsonArray;
```
You can also make this a `void` function (you don't use the returned value anyway).

AltNico replied:
I just realized I did not use that function at all ;P Thank you for the hint. Now down to 315 new lines.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+					e.printStackTrace();
+				} catch (JSONException e) {
+					e.printStackTrace();
+				}
```
you can collapse the two `catch` branches to

```
} catch (ParseException | JSONException e) {
```

as long as you do the same thing in both. (Java 7 feature)

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
```
Please remove the unused imports (IntelliJ can do this automatically with `Alt+Enter` -> Optimize Imports).

AltNico replied:
As I use Geany for programming (that's like Notepad on Windows with syntax highlighting), I cannot do this easily. Can you tell me which imports are unused or do it yourself after merging?

johan12345 replied:
Okay, I'll do that after merging.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+	private JSONArray sortByDate(JSONArray jsonArray) throws JSONException {
+		final JSONArray sortedJsonArray = new JSONArray();
+
+		final List<JSONObject> jsonValues = new ArrayList<JSONObject>();
```
You can just use `new ArrayList<>()` (Java 7 feature).

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+		final Substitution substitution = new Substitution();
+		// Set class
+		final String classId = change.getString("course");
+		if (!classId.equals("0") && !classId.equals(null)) {
```
Using `!classId.equals(null)` does not make sense. If `classId` is `null`, this will cause a `NullPointerException`. You probably meant `classId != null`.

johan12345 commented on parser/src/main/java/me/vertretungsplan/parser/LegionBoardParser.java:
```diff
@@ -0,0 +1,347 @@
+			if (type.equals("Vertretung") || !coveringTeacherId.equals("0")) {
+				substitution.setPreviousTeacher(teachersHashMap.get(teacherId));
+			}
+			else {
```
`}` and `else` should be put on the same line in Java (I think this way is more common in C#).

johan12345 commented:
I found no problems when testing the code, just some code style issues (see the inline comments).

AltNico commented:
@johan12345 Thank you for your feedback, I have fixed everything. See my other comments.
